tests/examples/shims/root.lua

```diff
@@ -1,9 +1,126 @@
 local root = {_tags={}}
 
+local hotkeys = nil
+
 function root:tags()
     return root._tags
 end
 
+function root:size() --TODO use the screens
+    return 0, 0
+end
+
+function root:size_mm()
+    return 0, 0
+end
+
+function root.cursor() end
+
+-- GLOBAL KEYBINDINGS --
+
+local keys = {}
+
+-- Note: the tests are not expected to create 1000 of them, so to keep the
+-- implementation minimal, it doesn't try to hash them and simply loop over
+-- every entry.
+function root.keys(k)
+    keys = k or keys
+    return keys
+end
+
+-- FAKE INPUTS --
+
+local conversion, mods = {
+    Super_L   = "Mod4",
+    Control_L = "Control",
+    Shift_L   = "Shift",
+    Alt_L     = "Mod1",
+    Super_R   = "Mod4",
+    Control_R = "Control",
+    Shift_R   = "Shift",
+    Alt_R     = "Mod1",
+}, {}
+
+local function add_modkey(key)
+    if not conversion[key] then return end
+    mods[conversion[key]] = true
+end
+
+local function get_mods()
+    local ret = {}
+
+    for mod in pairs(mods) do
+        table.insert(ret, mod)
+    end
+
+    return ret
+end
+
+local function remove_modkey(key)
+    if not conversion[key] then return end
+    mods[conversion[key]] = nil
+end
+
+local function match_modifiers(mods1, mods2)
+    if #mods1 ~= #mods2 then return false end
+
+    for _, mod1 in ipairs(mods1) do
+        local found = false
+        for _, mod2 in ipairs(mods2) do
+            if mod1 == mod2 then
+                found = true
+                break
+            end
+        end
+        if not found then return false end
+    end
+
+    return true
+end
+
+local function execute_press_keybindings(key, event)
+    -- It *could* be extracted from gears.object private API, but it's equally
+    -- ugly as using the list used by the hotkey widget.
+    if not hotkeys then
+        hotkeys = require("awful.key").hotkeys
+    end
```

Comment:
Huh?!? Shouldn't this be using `keys`, which is the list saved by `root.keys()`?

Reply:
Not currently, at least not cleanly. The callbacks are stored in the signal connections, so getting them is potentially going to run into "for some reason there is now many signals on "press" and I don't know which to use.

In any case, this is part of the reason I said in the header description that the way Awesome handle keys has to change as it is not extremely inflexible and race-prone. It doesn't matter as long as the user never change `root.keys()`, but as soon as you do, things get messy, fast.

Using `hotkeys` here is a workaround to avoid having to add even more code to fix the "real" issue. At some point, `capi.key` and `awul.key` should be merged into an unified object with a clean API. However it can't be done without breaking the API because an `awful.key` can have multiple `capi.key`.

Which bring us to the limitation of the current design and why `root.keys()` can't be used. @actionless `awful.hotkey_popup` ran into the same issue and added a parallel way to store the key properties. The shims now makes use of it instead of duplicating it. Also note that another commit add the `press` and `release` property to the previously mentioned parallel way to store key properties.